libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/JsonExtensions.cs

```diff
@@ -38,34 +38,40 @@ public static T Dequeue<T>(this List<T> queue)
         /// <param name="token">A JSON Token value which may have some binding paths.</param>
         /// <param name="state">A scope for looking up variables.</param>
         /// <returns>Deep data binding result.</returns>
-        public static JToken ReplaceJTokenRecursively(this JToken token, object state)
+        public static JToken ReplaceJToken(this JToken token, object state)
+        {
+            if (token.Type == JTokenType.String)
```

Comment:
Can you add a comment here saying
// If the token type is string, it's already been evaluated previously

Reply:
Also, token type is string probably not a good indicator that this value has been evaluated. Should be use value.expressionText is null?

Reply:
Or can we do this totally inside ValueExpression where we have all the information.  So that can save us to do this after the ValueExpression's evaluation is already over and we have to find a way to tell whether it's been evaluated or not.